File: GomokuAI/wrapper.py
from flask import Flask, request, jsonify
import subprocess
import threading
import queue

app = Flask(__name__)

process = subprocess.Popen(
    ['./build/pbrain-rapfi', '--mode', 'gomocup'],
    stdin=subprocess.PIPE,
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE,
    text=True
)

command_queue = queue.Queue()
response_queue = queue.Queue()

def process_io():
    while True:
        try:
            command = command_queue.get(timeout=0.1)
            process.stdin.write(command + '\n')
            process.stdin.flush()
            response = process.stdout.readline().strip()
            response_queue.put(response)
        except queue.Empty:
            continue

io_thread = threading.Thread(target=process_io, daemon=True)
io_thread.start()

@app.route('/test')
def test():
    return jsonify({"status": "Engine is running"})

@app.route('/command', methods=['POST'])
def command():
    cmd = request.json.get('command')
    if not cmd:
        return jsonify({"error": "No command provided"}), 400

    command_queue.put(cmd)

    try:
        response = response_queue.get(timeout=5)
        return jsonify({"response": response})
    except queue.Empty:
        return jsonify({"error": "Timeout waiting for response"}), 408

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5005)

File: GomokuAI/wrapper.js
const express = require('express');
const { spawn } = require('child_process');
const { Readable, Writable } = require('stream');

const app = express();
app.use(express.json());

const process = spawn('./build/pbrain-rapfi', ['--mode', 'gomocup'], {
    stdio: ['pipe', 'pipe', 'pipe']
});

const commandQueue = [];
const responseQueue = [];

const inputStream = new Writable({
    write(chunk, encoding, callback) {
        process.stdin.write(chunk, encoding, callback);
    }
});

const outputStream = new Readable({
    read() {}
});

process.stdout.on('data', (data) => {
    const response = data.toString().trim();
    if (responseQueue.length > 0) {
        const resolve = responseQueue.shift();
        resolve(response);
    }
});

process.on('error', (err) => {
    console.error('Failed to start subprocess:', err);
});

app.get('/test', (req, res) => {
    res.json({ status: "Engine is running" });
});

app.post('/command', async (req, res) => {
    const cmd = req.body.command;
    if (!cmd) {
        return res.status(400).json({ error: "No command provided" });
    }

    commandQueue.push(cmd);

    inputStream.write(cmd + '\n');

    try {
        const response = await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error("Timeout waiting for response")), 5000);
            responseQueue.push((data) => {
                clearTimeout(timeout);
                resolve(data);
            });
        });
        res.json({ response });
    } catch (err) {
        res.status(408).json({ error: err.message });
    }
});

const PORT = 5005;
app.listen(PORT, () => {
    console.log(`Server running on http://0.0.0.0:${PORT}`);
});
